File: 51.DFSAndBFSAlgo.py
# Depth First Search (DFS) 
# explores a graph by going as deep as possible along each branch before backtracking.

# Breadth First Search (BFS) explores all neighbors at the present depth 
# before moving on to nodes at the next depth level.

# Explanation

# Graphs are made up of nodes (vertices) and edges (connections).
# To traverse or explore all vertices in a graph:

# DFS uses a stack (or recursion).

# BFS uses a queue.

# They are used in:

# Finding connected components.

# Pathfinding.

# Cycle detection.

# Topological sorting, etc.

# Example Graph

# Let’s take the following undirected graph:

#      1
#     / \
#    2   3
#   / \
#  4   5

# Adjacency representation:

# graph = {
#   1: [2, 3],
#   2: [1, 4, 5],
#   3: [1],
#   4: [2],
#   5: [2]
# }

from collections import deque

graph = {
  1: [2, 3],
  2: [1, 4, 5],
  3: [1],
  4: [2],
  5: [2]
}

# DFS 
def dfs(start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start, end = " ")
    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs(neighbor, visited)

# BFS
def bfs(start):
    visited = set()
    q = deque([start])
    visited.add(start)
    while q:
        node = q.popleft()
        print(node, end = " ")
        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                q.append(neighbor)

print("DFS Traversal starting from the node 1:")
dfs(1)
print("\nBFS Traversal starting from the node 1:")
bfs(1)

File: 51.DFSAndBFSAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.util.*;

public class DFS_BFS_Graph {
    
    // Adjacency List Representation
    static Map<Integer, List<Integer>> graph = new HashMap<>();

    // ---------------- DFS Implementation ----------------
    static void dfs(int start, Set<Integer> visited) {
        visited.add(start);                          // Mark as visited
        System.out.print(start + " ");               // Print current node
        
        for (int neighbor : graph.get(start)) {      // Explore neighbors
            if (!visited.contains(neighbor)) {       // If not visited
                dfs(neighbor, visited);              // Recursive DFS
            }
        }
    }

    // ---------------- BFS Implementation ----------------
    static void bfs(int start) {
        Queue<Integer> q = new LinkedList<>();       // Queue for BFS
        Set<Integer> visited = new HashSet<>();      // Track visited nodes
        
        q.add(start);
        visited.add(start);
        
        while (!q.isEmpty()) {
            int node = q.poll();                     // Dequeue node
            System.out.print(node + " ");            // Print current node
            
            for (int neighbor : graph.get(node)) {   // Explore neighbors
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    q.add(neighbor);                 // Enqueue neighbor
                }
            }
        }
    }

    // ---------------- Main Method ----------------
    public static void main(String[] args) {
        // Build graph
        graph.put(1, Arrays.asList(2, 3));
        graph.put(2, Arrays.asList(1, 4, 5));
        graph.put(3, Arrays.asList(1));
        graph.put(4, Arrays.asList(2));
        graph.put(5, Arrays.asList(2));

        System.out.println("DFS Traversal starting from node 1:");
        dfs(1, new HashSet<>());                     // Start DFS
        
        System.out.println("\nBFS Traversal starting from node 1:");
        bfs(1);                                      // Start BFS
    }
}
